File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int max(int a,int b) {return a>b?a:b;}
    int lengthOfLongestSubstring(string s) {
        map<char,int> mp;
        int maxLen = 0,start=0;
        for(int i=0;i<s.length();i++) {
            if (mp.find(s[i]) != mp.end() && mp[s[i]]>=start){
                start = mp[s[i]]+1;
            }
            mp[s[i]] = i;
            maxLen = max(maxLen,i-start + 1);
        }
        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.rs
use std::collections::HashMap;
impl Solution {
    pub fn max(a:i32,b:i32)->i32 {
        if a>b {
            return a;
        }else {
            return b;
        }
    }
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut start=0;
        let mut maxLen=0;
        let mut mp:HashMap<char,usize> = HashMap::new();
        for (i,ele) in s.chars().enumerate() {
            if let Some(idx) = mp.insert(ele,i){
                if idx>=start {
                    start = idx+1;
                }
            }
            maxLen = Self::max(maxLen,(i-start+1) as i32);
        }
        maxLen
    }
}